pkg/fanal/analyzer/language/nodejs/yarn/yarn.go

```diff
@@ -169,11 +169,54 @@ func (a yarnAnalyzer) parsePackageJsonDependencies(fsys fs.FS, path string) (map
 	}
 	defer func() { _ = f.Close() }()
 
-	pkg, err := a.packageJsonParser.Parse(f)
+	rootPkg, err := a.packageJsonParser.Parse(f)
 	if err != nil {
 		return nil, xerrors.Errorf("parse error: %w", err)
 	}
 
 	// Merge dependencies and optionalDependencies
-	return lo.Assign(pkg.Dependencies, pkg.OptionalDependencies), nil
+	dependencies := lo.Assign(rootPkg.Dependencies, rootPkg.OptionalDependencies)
+
+	if rootPkg.Workspaces != nil {
+		pkgs, err := a.traverseWorkspaces(fsys, rootPkg.Workspaces)
+		if err != nil {
+			return nil, xerrors.Errorf("traverse workspaces error: %w", err)
+		}
+		for _, pkg := range pkgs {
+			dependencies = lo.Assign(dependencies, pkg.Dependencies, pkg.OptionalDependencies)
+		}
+	}
+
+	return dependencies, nil
+}
+
+func (a yarnAnalyzer) traverseWorkspaces(fsys fs.FS, workspaces []string) ([]packagejson.Package, error) {
+	var pkgs []packagejson.Package
+
+	required := func(path string, _ fs.DirEntry) bool {
+		return filepath.Base(path) == types.NpmPkg
+	}
+
+	walkDirFunc := func(path string, d fs.DirEntry, r dio.ReadSeekerAt) error {
+		pkg, err := a.packageJsonParser.Parse(r)
+		if err != nil {
+			return xerrors.Errorf("unable to parse %q: %w", path, err)
+		}
+		pkgs = append(pkgs, pkg)
+		return nil
+	}
+
+	for _, workspace := range workspaces {
```

Comment:
I'm thinking of 1 case - what about nested workspaces?
I mean the case when `rootPackage` -> `package1` -> `package2`.

Do you know about it? Is it possible?

Reply:
This is possible since `walkDir` takes the workspace path and we traverse through all child folders. I [added](https://github.com/aquasecurity/trivy/pull/4664/files/c2cc7234f074a73970ebd218bfa9cc57f8cdd02f#diff-2bb501f620e30a752277555f8ecc56c79563f19e16de1ff1ced30371332f0d72) this case to the tests.

Reply:
i mean when `app` includes `package1` and `package1` includes `package2`:
```
├── package1
│   ├── package2
│   │   └── package.json
│   └── package.json
├── package.json
└── yarn.lock
```
app:
```
➜  4664 cat package.json
{
   "name": "app",
   "version": "1.0.0",
   "main": "index.js",
   "license": "MIT",
   "private": true,
   "workspaces": [
      "package1"
   ]
}
```
package1:
```
➜  4664 cat ./package1/package.json 
➜  4664 cat ./package1/package.json
{
   "name":"package1",
   "version":"1.0.0",
   "main":"index.js",
   "license":"MIT",
   "private":true,
   "workspaces": [
     "package2"
   ]
}
```
package2:
```
➜  4664 cat ./package1/package2/package.json 
{
  "name": "package2",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "private": true,
  "dependencies": {
    "ms": "^2.1.3"
  }
}
➜  
```

But i have played around with these files and looks like `yarn` can't get dependencies from nested workspaces.
`npm` only contains workspaces, but no dependencies:
```
➜  4664 cat package-lock.json 
{
   "name": "app",
   "version": "1.0.0",
   "lockfileVersion": 3,
   "requires": true,
   "packages": {
      "": {
         "name": "app",
         "version": "1.0.0",
         "license": "MIT",
         "workspaces": [
            "package1"
         ]
      },
      "node_modules/package1": {
         "resolved": "package1",
         "link": true
      },
      "package1": {
         "version": "1.0.0",
         "license": "MIT",
         "workspaces": [
            "package2"
         ]
      }
   }
}

```
So let's use your solution. If there is a user who can do this, we will update the logic.

Reply:
As far as I know, only one package.json in the repository root can have the `workspaces` field

Reply:
In order for nested packages to be included in the workspace, wildcards are used. for example: `"workspaces": ["packages/**"]`